byok documentation (#6634)
camunda/camunda-docs: docs/components/saas/byok/aws-kms-setup.md

## Step 2: Create and configure a KMS key in AWS

Follow these steps in your AWS account:

1. Open the **AWS KMS console**.
2. Click **Create key**.
3. Choose **Symmetric** key type.
4. Select the **Region** matching your Camunda cluster.
5. Click **Next: Key administrators** and assign at least one administrator.
6. Click **Next: Key usage permissions**. Here, you need to grant your Camunda cluster AWS Role access.

### Example KMS key policy

Replace `<your-cluster-role-arn>` with the AWS Role ARN from Step 1:

```json
{
  "Sid": "Allow Camunda Cluster AWS Role basic key access",
  "Effect": "Allow",
  "Principal": {
    "AWS": "<your-cluster-role-arn>"
  },
  "Action": [
    "kms:DescribeKey",
    "kms:GenerateDataKey*",
    "kms:Decrypt"
  ],
  "Resource": "*"
},
{
  "Sid": "Allow Camunda Cluster AWS Role to create grants for provisioning encrypted EBS volumes",
  "Effect": "Allow",
  "Principal": {
    "AWS": "<your-cluster-role-arn>"
  },
  "Action": [
    "kms:CreateGrant"
  ],
  "Resource": "*"
}
```

:::warning Key policy warnings

- Ensure you do not restrict the Camunda Cluster AWS Role from any required KMS actions; otherwise encryption will fail.

> - Keys must remain in the same region as your cluster.

Reviewer: ```suggestion
- Keys must reside in the same region as your cluster.
```

- Any revocation of access will immediately break the Camunda cluster.

:::

7. Click **Finish** to create the key.
8. Copy the **Key ARN**; you will need it in the Camunda Console.

:::note Screenshot placeholder
[Insert screenshot of AWS KMS key details showing Key ARN]
:::

## Step 3: Associate KMS key with Camunda cluster

1. Return to the Camunda Console.
2. In the cluster details page, locate the KMS Key ARN input field.
3. Paste your AWS KMS Key ARN from Step 2.
4. Confirm and apply the key. The Camunda Operator will provision storage using your key for:
   1. Document handling storage
   2. Backup storage
   3. Orchestration Cluster persistent disks
   4. Elasticsearch persistent disks

:::note Screenshot placeholder
[Insert screenshot showing KMS Key ARN input in cluster details]
:::

## Step 4: Verify encryption and logging

- AWS CloudTrail automatically logs all KMS operations.
- Set up CloudWatch alerts for KMS key deletion or unauthorized access attempts.

> - Verify your cluster’s storage volumes are using the correct KMS key. This is possible, but not required.

Author: Added a bit, let me know what you think?

---

> id: aws-byok-setup

Reviewer 2: ```suggestion
id: aws-external-encryption-setup
```

title: "Encryption at rest using external encryption keys"
description: "Learn how to configure Camunda 8 SaaS using external AWS encryption keys."
---

> Learn how to configure Bring Your Own Key (BYOK) for Camunda 8 SaaS using AWS KMS.

Reviewer 2: ```suggestion
Learn how to configure external encryption for Camunda 8 SaaS using AWS KMS.
```

## Prerequisites

Before you begin, make sure you have:

| Requirement           | Description                                                                 |
| --------------------- | --------------------------------------------------------------------------- |
| AWS account           | Access to an AWS account with KMS permissions.                              |
| KMS permissions       | Ability to create and manage KMS keys and attach policies.                  |
| Cluster region        | KMS key must be created in the same AWS region as your Camunda 8.8 cluster. |
| Technical familiarity | Some experience with AWS console, IAM roles, and KMS is recommended.        |

:::warning Important

- Deleting or disabling your KMS key will make your cluster and data inaccessible.

> - BYOK keys cannot be rotated by Camunda 8.8; any rotation must be done manually in AWS KMS by the customer.

Reviewer 2: ```suggestion
- Key management is done in the AWS KMS by the customer. Camunda cannot rotate encryption keys.
```

> - Make sure the key policy grants your Camunda cluster the required access.

Reviewer 2: ```suggestion
```

This will be part of the steps. I'm not sure we need to call it out again.

:::

## Step 1: Create a Camunda 8 SaaS cluster

1. Sign in to the [Camunda Console](https://console.camunda.io/).
2. Navigate to the **Cluster** section and click **Create New Cluster**.
3. For **Region**, select an AWS region where you want your cluster deployed.
4. Under **Encryption at rest**, choose **External**.
5. Click **Create Cluster**.

:::note Screenshot placeholder
[Insert screenshot of cluster creation page with "External" encryption selected]
:::

> After creation, the Console will display a **unique Camunda Cluster IAM role ARN**. You will use this in the next step when configuring your KMS key.

Reviewer 2: ```suggestion
After creation, the Console will display a **AWS Role ARN** that is unique to your Camunda Cluster. You will use this in the next step when configuring your KMS key.
```

## Step 2: Create and configure a KMS key in AWS

Follow these steps in your AWS account:

1. Open the **AWS KMS console**.
2. Click **Create key**.
3. Choose **Symmetric** key type.
4. Select **Region** matching your Camunda cluster.
5. Click **Next: Key administrators** and assign at least one administrator.
6. Click **Next: Key usage permissions**. Here, you need to grant your Camunda cluster role access.

### Example KMS key policy

> Replace `<tenant-iam-role-arn>` with the IAM role ARN from Step 1:

Reviewer 2: ```suggestion
Replace `<your-cluster-role-arn>` with the AWS Role ARN from Step 1:
```

```json
{
  "Sid": "Allow Camunda tenant IAM Role basic key access",
  "Effect": "Allow",
  "Principal": {

>     "AWS": "<tenant-iam-role-arn>"

Reviewer 2: ```suggestion
    "AWS": "<your-cluster-role-arn>"
```

  },
  "Action": [
    "kms:DescribeKey",
    "kms:GenerateDataKey*",
    "kms:Decrypt"
  ],
  "Resource": "*"
},
{
  "Sid": "Allow Camunda tenant IAM Role to create grants for provisioning encrypted EBS volumes",
  "Effect": "Allow",
  "Principal": {

>     "AWS": "<tenant-iam-role-arn>"

Reviewer 2: ```suggestion
    "AWS": "<your-cluster-role-arn>"
```

  },
  "Action": [
    "kms:CreateGrant"
  ],
  "Resource": "*"
}
```

:::warning Key policy warnings

> - Ensure you do not restrict the Camunda cluster IAM role from any required KMS actions; otherwise encryption will fail.

Reviewer 2: ```suggestion
- Ensure you do not restrict the Camunda Cluster AWS Role from any required KMS actions; otherwise encryption will fail.
```

:::

7. Click **Finish** to create the key.
8. Copy the **Key ARN**; you will need it in the Camunda Console.

:::note Screenshot placeholder
[Insert screenshot of AWS KMS key details showing Key ARN]
:::

## Step 3: Associate KMS key with Camunda cluster

1. Return to the Camunda Console.
2. In the cluster details page, locate the KMS Key ARN input field.
3. Paste your AWS KMS Key ARN from Step 2.
4. Confirm and apply the key. The Camunda Operator will provision storage using your BYOK key for:
   1. Document handling storage
   2. Backup storage
   3. Orchestration Cluster persistent disks
   4. Elasticsearch persistent disks

:::note Screenshot placeholder
[Insert screenshot showing KMS Key ARN input in cluster details]
:::

## Step 4: Verify encryption and logging

- AWS CloudTrail automatically logs all KMS operations.
- Set up CloudWatch alerts for KMS key deletion or unauthorized access attempts.
- Verify your cluster’s storage volumes are using the correct KMS key. This is possible, but not required.

:::warning Monitoring reminder
You are responsible for monitoring your key usage and access logs.
:::

## Additional considerations

> - **Key rotation**: AWS supports manual rotation; Camunda 8.8 does not rotate customer keys.

Reviewer 3: When we say manual, we refer to external to Camunda. KMS rotation is not manual, as customers have automated procedures to rotate keys.
Using the manual term here may be confusing. Can we say something inline that Amazon KMS must perform key rotation as it owns the key lifecycle?

---
id: aws-external-encryption-setup
title: "Encryption at rest using external encryption keys"
description: "Learn how to configure Camunda 8 SaaS using external Amazon encryption keys."
---

Learn how to configure external encryption for Camunda 8 SaaS using **Amazon KMS**.

## Prerequisites

Before you begin, make sure you have:

| Requirement           | Description                                                                |
| --------------------- | -------------------------------------------------------------------------- |
| Amazon account        | Access to an Amazon account with KMS permissions.                          |
| KMS permissions       | Ability to create and manage KMS keys and attach policies.                 |
| Cluster region        | KMS key must reside in the same Amazon region as your Camunda 8.8 cluster. |
| Technical familiarity | Some experience with Amazon console, IAM roles, and KMS is recommended.    |

:::warning Important

> 

Reviewer 4: ```suggestion
```

- Deleting or disabling your KMS key will make your cluster and data inaccessible.
- Key management is performed fully on the customer side in Amazon KMS. Camunda cannot rotate your encryption keys.

> :::

Reviewer 4: ```suggestion
```

## Step 1: Create a Camunda 8 SaaS cluster

1. Sign in to the [Camunda Console](https://console.camunda.io/).
2. Navigate to the **Cluster** section and click **Create New Cluster**.
3. For **Region**, select an Amazon region where you want your cluster deployed.

> 4. Under **Encryption at rest**, choose **External**.

Reviewer 4: Before this, the customer has the option to choose if they want single or dual region. This also affects the key encryption

5. Click **Create Cluster**.

> :::note Screenshot placeholder

Reviewer 4: ```suggestion
[!NOTE] Screenshot placeholder
```

[Insert screenshot of cluster creation page with "External" encryption selected]
:::

After creation, the Console will display an **Amazon Role ARN** that is unique to your Camunda cluster. You will use this in the next step when configuring your KMS key.

## Step 2: Create and configure a KMS key in Amazon

Follow these steps in your Amazon account:

1. Open the **Amazon KMS console**.
2. Click **Create key**.
3. Choose **Symmetric** key type.
4. Select the **Region** matching your Camunda cluster.
5. Click **Next: Key administrators** and assign at least one administrator.
6. Click **Next: Key usage permissions**. Here, you need to grant your Camunda cluster Amazon Role access.

> 

Reviewer 4: In the case, the customer chooses dual region, they should create the key as multi-region key

### Example KMS key policy

Replace `<your-cluster-role-arn>` with the Amazon Role ARN from Step 1:

```json
{

>   "Sid": "Allow Camunda Cluster Amazon Role basic key access",

Reviewer 4: ```suggestion
  "Sid": "Allow Camunda tenant IAM Role basic key access",
```

  "Effect": "Allow",
  "Principal": {
    "AWS": "<your-cluster-role-arn>"
  },
  "Action": [
    "kms:DescribeKey",
    "kms:GenerateDataKey*",
    "kms:Decrypt"
  ],
  "Resource": "*"
},
{

>   "Sid": "Allow Camunda Cluster Amazon Role to create grants for provisioning encrypted EBS volumes",

Reviewer 4: ```suggestion
  "Sid": "Allow Camunda tenant IAM Role to create grants for provisioning encrypted EBS volumes",
```

  "Effect": "Allow",
  "Principal": {
    "AWS": "<your-cluster-role-arn>"
  },
  "Action": [
    "kms:CreateGrant"
  ],
  "Resource": "*"
}

> ```

Reviewer 4: ```suggestion
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "Enable IAM User Permissions",
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::<customer-aws-account>:root"
      },
      "Action": "kms:*",
      "Resource": "*"
    },
    {
      "Sid": "Allow Camunda tenant IAM Role basic key access",
      "Effect": "Allow",
      "Principal": {
        "AWS": "<tenant-role-arn>"
      },
      "Action": [
        "kms:DescribeKey",
        "kms:GenerateDataKey*",
        "kms:Decrypt"
      ],
      "Resource": "*"
    },
    {
      "Sid": "Allow Camunda tenant IAM Role to create grants for provisioning encrypted EBS volumes",
      "Effect": "Allow",
      "Principal": {
        "AWS": "<tenant-role-arn>"
      },
      "Action": [
        "kms:CreateGrant",
        "kms:ListGrants"
       ]
      "Resource": "*"
    }
  ]
}
```

Reviewer 2: Updated this to contain also RevokeGrant + the permissions for replication of backups (ReEncrypt*)

:::warning Key policy warnings

> 

Reviewer 4: ```suggestion
```

- Ensure you do not restrict the Camunda Cluster Amazon Role from any required KMS actions; otherwise encryption will fail.

> - Keys must reside in the same region as your cluster.

Reviewer 4: For dual region, this must be specified for dual region i.e. the replicated key should be on the same secondary region as the cluster

- Any revocation of access will immediately break the Camunda cluster.

> 

Reviewer 4: Additional info:
- After the key is created, its core details cannot be modified. It can always be replicated in other regions at any point

---
id: aws-external-encryption-setup
title: "Encryption at rest using external encryption keys"
description: "Learn how to configure Camunda 8 SaaS using external Amazon encryption keys."
---

Learn how to configure external encryption for Camunda 8 SaaS using **Amazon KMS**.

## Prerequisites

Before you begin, make sure you have:

| Requirement           | Description                                                                |
| --------------------- | -------------------------------------------------------------------------- |
| Amazon account        | Access to an Amazon account with KMS permissions.                          |
| KMS permissions       | Ability to create and manage KMS keys and attach policies.                 |
| Cluster region        | KMS key must reside in the same Amazon region as your Camunda 8.8 cluster. |
| Technical familiarity | Some experience with Amazon console, IAM roles, and KMS is recommended.    |

:::warning Important
- Deleting or disabling your KMS key will make your cluster and data inaccessible.

> - Key management is performed fully on the customer side in Amazon KMS. Camunda cannot rotate your encryption keys.

Reviewer 2: Customer can rotate the key. Feel free to try it in dev.

> - Key management is performed fully on the customer side in Amazon KMS. Camunda cannot rotate your encryption keys.

Reviewer 5: @Sijoma Right, but the storage itself will still be encrypted with the old version of the key, correct? At least that should be mentioned if that is the case

> - Key management is performed fully on the customer side in Amazon KMS. Camunda cannot rotate your encryption keys.

Reviewer 6: @sourcehawk If I remember right about the project planning meetings, the storage will be encrypted with the same key material until the moment the key gets rotated, which then will generate new key material. New data will be encrypted with the newest key material. When data needs to be decrypted (using the same KMS key), it will automatically detect which key material was used to encrypt it and use it to decrypt it. If we wanted the previous data to be encrypted with the new key material, a reencrypt operation would need to be done (https://docs.aws.amazon.com/kms/latest/developerguide/rotate-keys.html)

Its mentioned here
https://github.com/camunda/camunda-docs/pull/6634/files#diff-0e885e2ced806af95dcb04d3ad321df1100a56c0fae360816a6ce1d7cd255561R17

:::

## Step 1: Create a Camunda 8 SaaS cluster

1. Sign in to the [Camunda Console](https://console.camunda.io/).
2. Navigate to the **Cluster** section and click **Create New Cluster**.
3. For **Region**, select an Amazon region where you want your cluster deployed.
4. If desired, choose **Single region** or **Dual region**.  
   - For **Dual region**, your key setup will require multi-Region keys in AWS.
5. Under **Encryption at rest**, choose **External**.
6. Click **Create cluster**.

:::note
[Insert screenshot of cluster creation page with "External" encryption selected]
:::

After creation, the Console displays an **Amazon Role ARN** unique to your Camunda cluster. You will use this in the next step when configuring your KMS key.

## Step 2: Create and configure a KMS key in Amazon

Follow these steps in your Amazon account:

1. Open the **Amazon KMS console**.
2. Click **Create key**.
3. Choose **Symmetric** key type.
4. Select the **Region** matching your Camunda cluster.  

>    - For **Dual region**, create the key as a **multi-Region** key and ensure replication in the same secondary region as the cluster.

Reviewer 5: Feel free to resolve this, just an open question

5. Click **Next: Key administrators** and assign at least one administrator.
6. Click **Next: Key usage permissions**. Grant your Camunda cluster Amazon Role access.

### Example KMS key policy

Replace `<your-cluster-role-arn>` with the Amazon Role ARN from Step 1:

> 

Reviewer 4: ```suggestion
Replace <customer-aws-account> with your aws account id (the account where the kms key is created)
```

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "Enable IAM user permissions",
      "Effect": "Allow",
      "Principal": {
        "AWS": "arn:aws:iam::<customer-aws-account>:root"
      },
      "Action": "kms:*",
      "Resource": "*"
    },
    {
      "Sid": "Allow Camunda tenant IAM Role basic key access",
      "Effect": "Allow",
      "Principal": {
        "AWS": "<tenant-role-arn>"
      },
      "Action": [
        "kms:Encrypt",
        "kms:Decrypt",
        "kms:ReEncrypt*",
        "kms:DescribeKey",
        "kms:GenerateDataKey*"
      ],
      "Resource": "*"
    },
    {
      "Sid": "Allow Camunda tenant IAM Role to create grants for provisioning encrypted EBS volumes",
      "Effect": "Allow",
      "Principal": {
        "AWS": "<tenant-role-arn>"
      },
      "Action": [
        "kms:CreateGrant",
        "kms:ListGrants",
        "kms:RevokeGrant"
      ],
      "Resource": "*"
    }
  ]
}
```

> :::warning Key policy guidance

Reviewer 5: Something should be mentioned of key rotation if we cannot support that

- Ensure you do not restrict the Camunda cluster Amazon Role from required KMS actions; otherwise, encryption will fail.
- For dual region clusters, replicate keys in the correct secondary region.

> - After a key is created, core details cannot be modified. However, replication in additional regions is always possible later.

Reviewer 5: > However, replication in additional regions is always possible later.

What does this mean? Can a single region cluster become a dual region cluster?

- Any revocation of access immediately breaks the Camunda cluster.
:::

7. Click **Finish** to create the key.

> 8. Copy the **Key ARN**; you will need it in the Camunda Console.

Reviewer 5: ```suggestion
8. Copy the **KMS Key ARN**; you will need it in the Camunda Console.
```

:::note

> [Insert screenshot of Amazon KMS key details showing Key ARN]

Reviewer 5: ```suggestion
[Insert screenshot of AWS KMS key details showing Key ARN]
```

:::

## Step 3: Associate KMS key with Camunda cluster

1. Return to the Camunda Console.
2. On the cluster details page, locate the **KMS Key ARN** input field.  
   - For **Dual region**, two fields will be available—enter the correct key for each region.

> 3. Paste your Amazon KMS Key ARN(s) from Step 2.

Reviewer 5: ```suggestion
3. Paste your AWS KMS Key ARN(s) from Step 2.
```

4. Confirm and apply the key. The Camunda Operator provisions storage using your key for:
   - Document handling storage
   - Backup storage
   - Orchestration cluster persistent disks
   - Elasticsearch persistent disks

> :::note

Reviewer 5: Do we need a note about the fact that it is not possible to edit the key after it has been applied, even if it failed to encrypt / key was invalid or did not exist?

[Insert screenshot showing KMS Key ARN input in cluster details]
:::

## Step 4: Verify encryption and logging

> After associating your Amazon KMS key with the Camunda cluster, follow these steps to ensure encryption is correctly applied and operations are logged.

Reviewer 5: ```suggestion
After associating your AWS KMS key with the Camunda cluster, follow these steps to ensure encryption is correctly applied and operations are logged.
```

### Verify encryption

- In the Camunda Console, check the cluster details page to confirm that the **KMS Key ARN** is correctly applied.
- Verify that all cluster storage volumes are encrypted with the configured key:
  - Document handling storage
  - Backup storage
  - Orchestration cluster persistent disks
  - Elasticsearch persistent disks

> - If needed, check the encryption status directly in the underlying storage service (e.g., Amazon EBS, S3, or Elasticsearch).

Reviewer 5: ```suggestion
- If needed, check the encryption status directly in the underlying storage service (e.g., AWS EBS, S3, or Elasticsearch).
```

### Monitor KMS usage

> - **Amazon CloudTrail** automatically logs all KMS operations, including key usage (`Encrypt`, `Decrypt`, `GenerateDataKey`).

Reviewer 5: ```suggestion
- **AWS CloudTrail** automatically logs all KMS operations, including key usage (`Encrypt`, `Decrypt`, `GenerateDataKey`).
```

> - Use **Amazon CloudWatch** to create alarms for:

Reviewer 5: ```suggestion
- Use **AWS CloudWatch** to create alarms for:
```

  - Key deletion or disabling
  - Unauthorized access attempts
- Regularly review logs to ensure your cluster uses the correct KMS key and no unauthorized operations occur.

:::warning Monitoring reminder
You are responsible for monitoring key usage and access logs. Use CloudTrail and CloudWatch to detect misconfigurations or unauthorized access.
:::

## Additional considerations

> - **Key rotation**: Amazon KMS owns the key lifecycle. Customers can enable [automatic rotation](https://docs.aws.amazon.com/kms/latest/developerguide/rotate-keys.html) or manage rotation procedures externally. Camunda does not control this process.

Reviewer 5: ```suggestion
- **Key rotation**: AWS KMS owns the key lifecycle. Customers can enable [automatic rotation](https://docs.aws.amazon.com/kms/latest/developerguide/rotate-keys.html) or manage rotation procedures externally. Camunda does not control this process.
```

> - **Cost implications**: Using Amazon KMS keys incurs storage and management charges in your Amazon account. See the [Camunda pricing model](./cost-and-troubleshooting.md) for details.

Reviewer 5: ```suggestion
- **Cost implications**: Using AWS KMS keys incurs storage and management charges in your Amazon account. See the [Camunda pricing model](./cost-and-troubleshooting.md) for details.
```

- **Failure scenarios**: If the key is deleted or permissions revoked, cluster data becomes inaccessible. See [troubleshooting steps](./cost-and-troubleshooting.md) for guidance on resolving access issues.

:::note Reference

> For more information, see the [Amazon KMS documentation](https://docs.aws.amazon.com/kms/latest/developerguide/overview.html).

Reviewer 5: ```suggestion
For more information, see the [AWS KMS documentation](https://docs.aws.amazon.com/kms/latest/developerguide/overview.html).
```

| Requirement           | Description                                                                |
| --------------------- | -------------------------------------------------------------------------- |
| Amazon account        | Access to an Amazon account with KMS permissions.                          |
| KMS permissions       | Ability to create and manage KMS keys and attach policies.                 |
| Cluster region        | KMS key must reside in the same Amazon region as your Camunda 8.8 cluster. |
| Technical familiarity | Some experience with Amazon console, IAM roles, and KMS is recommended.    |

:::warning Important

- Deleting or disabling your KMS key will make your cluster and data inaccessible.
- Key management is performed fully on the customer side in Amazon KMS. Camunda cannot rotate your encryption keys.
  :::

## Step 1: Create a Camunda 8 SaaS cluster

1. Sign in to the [Camunda Console](https://console.camunda.io/).
2. Navigate to the **Cluster** section and click **Create New Cluster**.
3. For **Region**, select an Amazon region where you want your cluster deployed.
4. If desired, choose **Single region** or **Dual region**.

>    - For **Dual region**, you must create one key per region. The keys do not need to be replicated; they can be separate keys as long as they are in the correct regions.

Reviewer 3: Lets clarify 
For **Dual region**
For clusters using **Dual region** backup

   - Document handling storage
   - Backup storage
   - Orchestration cluster persistent disks
   - Elasticsearch persistent disks

:::note
Once a key is applied, it cannot be edited or replaced—even if the key was invalid or encryption failed. To change keys, you must create a new cluster.
:::

:::note
[Insert screenshot showing KMS Key ARN input in cluster details]
:::

## Step 4: Verify encryption and logging

After associating your Amazon KMS key with the Camunda cluster, verify that encryption is working as expected.

### Verify encryption

- In the Camunda Console, check the cluster details page to confirm that the **KMS Key ARN** is correctly applied.
- Verify that all cluster storage volumes are encrypted with the configured key:
  - Document handling storage
  - Backup storage
  - Orchestration cluster persistent disks
  - Elasticsearch persistent disks
- If needed, check the encryption status directly in the underlying storage service (e.g., Amazon EBS, S3, or Elasticsearch).

### Monitor KMS usage

- **Amazon CloudTrail** automatically logs all KMS operations, including key usage (`Encrypt`, `Decrypt`, `GenerateDataKey`).

> - Use **Amazon CloudWatch** to create alarms for:

Reviewer 3: Could we expand to explain that alarms should be set up in a customer's AWs account, their KMS service.

## Step 2: Create and configure a KMS key in Amazon

Follow these steps in your Amazon account:

1. Open the **Amazon KMS console**.
2. Click **Create key**.
3. Choose **Symmetric** key type.
4. Select the **Region** matching your Camunda cluster.  
   - For **Dual region**, create the key as a **multi-Region** key and ensure replication in the same secondary region as the cluster.
5. Click **Next: Key administrators** and assign at least one administrator.
6. Click **Next: Key usage permissions**. Grant your Camunda cluster Amazon Role access.

### Example KMS key policy

Replace `<your-cluster-role-arn>` with the Amazon Role ARN from Step 1:

```json
{
  "Sid": "Allow Camunda Cluster Amazon Role basic key access",
  "Effect": "Allow",
  "Principal": {
    "AWS": "<your-cluster-role-arn>"
  },
  "Action": [
    "kms:DescribeKey",
    "kms:GenerateDataKey*",
    "kms:Decrypt"
  ],
  "Resource": "*"
},
{
  "Sid": "Allow Camunda Cluster Amazon Role to create grants for provisioning encrypted EBS volumes",
  "Effect": "Allow",
  "Principal": {
    "AWS": "<your-cluster-role-arn>"
  },
  "Action": [
    "kms:CreateGrant"
  ],
  "Resource": "*"

> }

Reviewer 2: I'm not sure this is enough - didnt we also need at least the ListGrants permissions for the EKS AutoMode? 

I would also suggest to already get the RevokeGrant so that we have future flexibility. 

Are we sure we don't need `kms:Encrypt` & `kms:ReEncrypt*`? This was what was tested by @psetzer-camunda .

Related slack thread:
https://camunda.slack.com/archives/C0876PB1PK4/p1758549296797509?thread_ts=1758099969.470859&cid=C0876PB1PK4

   - Full support for dual-region encryption is under discussion; confirm with your Camunda contact.
5. Under **Encryption at rest**, choose **External**.
6. Click **Create cluster**.

After creation, note the **Amazon Role ARN** displayed in the Console for your cluster.

## Step 2: Create and configure an AWS KMS key

You can create the key either via CLI or manually in the AWS console.

### Option A: Create the key using AWS CLI

We provide automated scripts to create the necessary KMS key(s) with the correct policy and permissions. Choose the option that matches your backup configuration.

#### Single-region backup

Use this script to create a single KMS key in `eu-west-1` (Ireland).

**What the script does:**

- Creates a KMS key with the required policy for Camunda access.
- Sets up an alias for easier key management.
- Outputs the key ARN to provide to Camunda.

**Instructions:**

1. Download [create-byok-kms-key-single-region.sh](https://github.com/camunda/camunda-docs/tree/main/docs/components/saas/byok/downloads/create-byok-kms-key-single-region.sh).
2. Modify the following values at the top of the script:
   - `AWS_ACCESS_KEY_ID`
   - `AWS_SECRET_ACCESS_KEY`

>    - `AWS_SESSION_TOKEN` (if using temporary credentials)

Reviewer 4: ```suggestion
   - `AWS_SESSION_TOKEN` (if using temporary credentials)
   - `REGION` (replace with the region in which the cluster is created)
```

   - `TENANT_ROLE_ARN` (from the Camunda Console)
   - `YOUR_ACCOUNT_ID`
   - `ALIAS_NAME` (optional)
3. Make the script executable and run it.
4. Copy the outputted key ARN and provide it to Camunda.

#### Dual-region backup

Use this script to create a multi-region primary key in `eu-west-1` (Ireland) and a replica key in `eu-west-2` (London).

**What the script does:**

- Creates a multi-region primary key and replica key.
- Applies the correct policies to both keys.
- Outputs both key ARNs to provide to Camunda.

**Instructions:**

1. Download [create-byok-kms-key-multi-region.sh](https://github.com/camunda/camunda-docs/tree/main/docs/components/saas/byok/downloads/create-byok-kms-key-multi-region.sh).
2. Modify the same variables as above.
3. Make the script executable and run it.
4. Copy the two outputted key ARNs and provide them to Camunda.

:::note Alternative

> For dual-region setups, you can also run the single-region script twice—once in `eu-west-1` and once in `eu-west-2`. Make sure to modify the `REGION` variable before creating the second key.

Reviewer 4: ```suggestion
For dual-region setups, you can also run the single-region script twice—once in  cluster's region and once in backup region. Make sure to modify the `REGION` variable before creating the second key.
```

   - Navigate to the KMS service and select the correct region.
2. **Create a customer managed key**
   - Click **Create key**.
   - Choose **Symmetric** and **Encrypt and decrypt** usage.
3. **Add labels**
   - Add an alias (e.g., `camunda-saas-byok`).
   - Add a description (e.g., `KMS key for Camunda SaaS BYOK`).
4. **Define key administrators**
   - Select IAM users/roles that will administer the key.
5. **Define key usage permissions**
   - Skip this step; permissions are configured in the next step.
6. **Edit key policy**
   - Switch to policy view and replace the existing policy with the [provided key policy](https://github.com/camunda/camunda-docs/tree/main/docs/components/saas/byok/downloads/aws-kms-key-policy.json).
   - Replace `<YOUR_AWS_ACCOUNT_ID>` and `<TENANT_ROLE_ARN>` with your values.
7. **Finish and copy the ARN**
   - Click **Finish** and copy the key ARN to use in the Camunda Console.

#### Dual-region backup

You can either create a multi-region key and replica or create two single-region keys.

##### Method A: Multi-region key (recommended)

1. Follow the single-region steps, selecting **Multi-Region key** under **Advanced options**.

> 2. After creating the primary key in `eu-west-1`, go to **Regional replicas** and click **Create replica key**.

Reviewer 4: ```suggestion
2. After creating the primary key in cluster's region, go to **Regional replicas** and click **Create replica key**.
```

> 3. Select `eu-west-2` for the replica and confirm.

Reviewer 4: ```suggestion
3. Select the region for the replica and confirm. The region should be the same as the backup region.
```
